File: set1/challenge4.py
# http://cryptopals.com/sets/1/challenges/4/
from itertools import imap, izip_longest
from operator import itemgetter

# http://en.wikipedia.org/wiki/Etaoin_shrdlu
english_letter_frequency_lower = {
    'e': 13,
    't': 12,
    'a': 11,
    'o': 10,
    'i': 9,
    'n': 8,
    ' ': 7,
    's': 6,
    'r': 5,
    'h': 4,
    'd': 3,
    'l': 2,
    'u': 1,
}

english_letter_frequency_upper = {l.upper(): f for l, f in english_letter_frequency_lower.iteritems()}

english_letter_frequency = dict(english_letter_frequency_lower)
english_letter_frequency.update(english_letter_frequency_upper)

frequencies = [english_letter_frequency.get(chr(c), 0) for c in range(0, 256)]

def hex2int(chars):
    return int(''.join(chars), 16)


def hexdecode(input):
    args = [iter(input)] * 2
    hexdigits = izip_longest(fillvalue='', *args)
    return imap(hex2int, hexdigits)


def xor(data, key):
    return [a ^ b for a, b in izip_longest(data, [key], fillvalue=key)]


def stringify(ints):
    return ''.join([chr(i) for i in ints])


def score(candidate):
    result = sum([frequencies[c] for c in candidate])
    return (result, candidate)


def find_solution(candidates):
    sorted_candidates = sorted(candidates, key=itemgetter(0), reverse=True)
    best_score, solution = sorted_candidates[0]
    return solution


def decrypt(input):
    decoded = list(hexdecode(input))
    candidates = [score(xor(decoded, key)) for key in range(0, 256)]
    return find_solution(candidates)


def find(filename):
    with open(filename, 'r') as f:
        candidates = [score(decrypt(line[:-1])) for line in f.readlines()]
        return find_solution(candidates)


if __name__ == '__main__':
    input = '4.txt'
    print(stringify(find(input))),

File: set1/challenge4.hs
-- http://cryptopals.com/sets/1/challenges/4/
import Data.Bits (xor)
import Data.Char (chr, toLower)
import Data.Function (on)
import Data.List (sortBy)
import Numeric (readHex)
import System.IO (readFile)

groupBy :: Int -> [a] -> [[a]]
groupBy n [] = []
groupBy n xs = [take n xs] ++ groupBy n (drop n xs)

hex2int :: [Char] -> Int
hex2int x = fst  $ head  $ readHex x

hexDecode :: [Char] -> [Int]
hexDecode input = [hex2int i | i <- groupBy 2 input]

xorList :: [Int] -> Int -> [Int]
xorList list key = (zipWith xor list (repeat key))

frequency :: Int -> Int
frequency byte
    | char == 'e' = 13
    | char == 't' = 12
    | char == 'a' = 11
    | char == 'o' = 10
    | char == 'i' = 9
    | char == 'n' = 8
    | char == ' ' = 7
    | char == 's' = 6
    | char == 'r' = 5
    | char == 'h' = 4
    | char == 'd' = 3
    | char == 'l' = 2
    | char == 'u' = 1
    | otherwise   = 0
    where char = toLower $ chr byte

score :: [Int] -> (Int, [Int])
score candidate =
    let result = sum [frequency c | c <- candidate]
    in (result, candidate)

stringify :: [Int] -> [Char]
stringify list = [chr d | d <- list]

findSolution :: [(Int, [Int])] -> [Int]
findSolution candidates =
    let sortedCandidates = sortBy (flip (compare `on` fst)) candidates
    in snd $ head sortedCandidates

decrypt :: [Char] -> [Int]
decrypt input =
    let decoded = hexDecode input
        candidates = [score $ xorList decoded key | key <- [0..255]]
    in findSolution candidates

find :: [[Char]] -> [Int]
find lines =
    let candidates = [score $ decrypt line | line <- lines]
    in findSolution candidates

main = do
    contents <- readFile "4.txt"
    putStr $ stringify $ find $ lines contents
